File: 2021/python/141.flatten-binary-tree-2-linkedlist.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        curr = root
        while curr:
            if curr.left:
                rightmost = curr.left
                while rightmost.right:
                    rightmost = rightmost.right
                rightmost.right = curr.right
                curr.right = curr.left
                curr.left = None
            
            curr = curr.right

File: 2021/141.flatten-binary-tree-2-linkedlist.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func flatten(root *TreeNode)  {
    if root == nil {
        return
    }
    
//     var curr *TreeNode
//     curr = root
    curr := root
    for curr != nil {
        if curr.Left != nil {
//             var rightmost *TreeNode
//             rightmost = curr.Left
            rightmost := curr.left
            for rightmost.Right != nil {
                rightmost = rightmost.Right
            }
            rightmost.Right = curr.Right
            curr.Right = curr.Left
            curr.Left = nil
        }

        curr = curr.Right
    }  
}
